### PR TITLE
catalog-client: allow `filter` param to be specified multiple times

### PR DESCRIPTION
**Pull request by WillSewell**
## Hey, I just made a Pull Request!

The backend permits the `filter` param to be specified multiple times

https://github.com/backstage/backstage/blob/f9d077f5f6c3bf1e20255359798dd1febbeb3e27/plugins/catalog-backend/src/database/types.ts#L129-L131

https://github.com/backstage/backstage/blob/f9d077f5f6c3bf1e20255359798dd1febbeb3e27/plugins/catalog-backend/src/service/request/parseEntityFilterParams.ts#L25-L42

this adds support for this to the client.

Currently this introduces a _breaking change_ because it changes the type of the public `filter` field of `CatalogEntitiesRequest`. I'm not sure what your philosophy is on introducing breaking changes. If you would prefer I didn't, I could adjust the type to `Record<string, string | string[]>[] | Record<string, string | string[]> | undefined` at the cost of additional complexity.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. (_pending decision on breaking change_)
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/catalog-client/src/types.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@
 import { Entity, EntityName, Location } from '@backstage/catalog-model';
 
 export type CatalogEntitiesRequest = {
-  filter?: Record<string, string | string[]> | undefined;
+  filter?: Record<string, string | string[]>[] | undefined;
```
Comment (freben) [inline packages/catalog-client/src/types.ts:20]: you probably want to make this support both the old type and the array form of the old type (and undefined). That way, callers don't need to change everywhere. Then your actual code can check `Array.isArray(filter) ? filter : [filter]` or do `[filter].flat()` or so, to easily handle both variants.

**Review by freben — COMMENTED** [review]
This needs a changeset too!

**Comment by WillSewell** [comment]
Thanks for the feedback. I picked a "minor" update rather than a "patch" for the changeset because I see this as adding functionality, rather than a bugfix.

File: packages/catalog-client/src/types.ts [hunk 2]
```diff
@@ -17,7 +17,10 @@
 import { Entity, EntityName, Location } from '@backstage/catalog-model';
 
 export type CatalogEntitiesRequest = {
-  filter?: Record<string, string | string[]> | undefined;
+  filter?:
+    | Record<string, string | string[]>[]
+    | Record<string, string | string[]> // Legacy type preserved for backwards compatibility
```
Comment (Rugvip) [inline packages/catalog-client/src/types.ts:22]: Could skip this comment tbh, it's there for convenience to not have to always wrap in an array

File: .changeset/twenty-peas-deny.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/catalog-client': minor
```
Comment (Rugvip) [inline .changeset/twenty-peas-deny.md:2]: Let's make this a `patch`, we use that for non-breaking changes while we're on 0.x

**Review by Rugvip — COMMENTED** [review]
Thanks, looks good! 👍 Just a few nits

File: .changeset/twenty-peas-deny.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/catalog-client': patch
+---
+
+Allow `filter` param to be specified multiple times
```
Comment (Rugvip) [inline .changeset/twenty-peas-deny.md:5]: Just spell out "parameter" to make CI happy :grin:

**Review by Rugvip — COMMENTED** [review]
Realized I didn't answer the initial API stability question, which we got here: https://backstage.io/docs/overview/stability-index

We still try to no break things unnecessarily though.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5493-9opdrdlvhbb8cdbti4uqqii7uu4kl70i.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608a9a528e7af6a2d2cbe759

**Review by freben — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]
:shipit:

**Comment by freben** [comment]
@WillSewell You got caught now in a very new feature - the build system tells you that you also need to update the public API documentation for this surface!

from the build log:

```
*************************************************************************************
* You have uncommitted changes to the public API of a package.                      *
* To solve this, run `yarn build:api-reports` and commit all api-report.md changes. *
*************************************************************************************

The conflicting file is catalog-client/api-report.md, with the following content:
```

would you like to try that out in that package and adding it to the pull request, and letting us know how it went?

**Comment by WillSewell** [comment]
Thanks for the tip. That check seems useful!

I think it went well - I just had to pull in the latest commits commits from `master` in order to get the required `@microsoft/api-extractor` dependency.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5493-9opdrdlvhbb8cdbti4uqqii7uu4kl70i.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608a9a528e7af6a2d2cbe759

**Comment by freben** [comment]
That did it!

**Pull request merged**